src/ability/crud.rs

```diff
@@ -0,0 +1,134 @@
+use crate::{promise::Promise, prove::TryProve};
+use std::{collections::BTreeMap, fmt::Debug};
+use url::Url;
+
+// FIXME move to promise.rs
+#[derive(Debug, Clone, PartialEq)]
+pub enum Field<T>
+where
+    T: Debug + Clone + PartialEq,
+{
+    Value(T),
+    Await(Promise), // FIXME
+}
+
+// FIXME macro to derive promise versions & delagted builder versions
+// ... also maybe Ipld
+
+pub struct Crud {
+    uri: Field<Url>,
+}
+
+pub struct CrudRead {
+    pub uri: Field<Url>,
+}
+
+pub struct CrudMutate {
+    uri: Field<Url>,
+}
+
+pub struct CrudCreate {
+    pub uri: Field<Url>,
+    pub args: BTreeMap<Box<str>, Field<String>>,
+}
+
+pub struct CrudUpdate {
+    pub uri: Field<Url>,
+    pub args: BTreeMap<Box<str>, Field<String>>,
+}
+
+pub struct CrudDestroy {
+    pub uri: Field<Url>,
+}
```

Comment:
This is setting a really high bar for type safety, which I commend, but could possibly be hard to realize.

I'm trying to categorize the 'needs' for these different types:
- in Delegations: Needs the ability to "name" the ability (kind of like a function name), and needs to know about which abilities can be derived from others. E.g. `account/info` can be derived from `account/noncritical`, but `account/delete` cannot. No need to know about arguments yet (unless we try to integrate conditions here, too).
- in Invocations: Needs not only the ability name, but also values for all arguments. Which abilities derive others is irrelevant. Stuff like `account/noncritical` doesn't even have a defined set of arguments and shouldn't even be usable in an invocation.
- in Workflows/Tasks(? I'm fuzzy on the terminology here): Needs ability arguments to have a representation for "deferred". Otherwise fairly similar needs to "in Invocations".
- in Receipts: Needs all arguments to be there, none of them optional nor deferred.

Maybe for now, we'll separate these into let's say two? I.e. require uses to define two structs for each , one for use within `Delegation` and one for use within `Invocation`/Tasks/Receipts.

I can see a future where we improve the situation with a derive proc-macro like in the future then.

Reply:
Yep, this is the "I wish I had HKTs" stuff that we talked about. You'll see the expanded code hopefully later today, but we can also macro-ize it. If it turns out to be too much of a pain, we can drop to a dynamic version, but so far writing this feels more straightforward than perhaps it feels?

This is also all standard lib — I think 99% of of people will just use the stuff off the shelf because they're general enough for _most_ use cases.

But hey, we'll see as this comes together :)